npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb

```diff
@@ -85,25 +86,43 @@ def fetch_files
 
       # If every entry in the lockfile uses the same registry, we can infer
       # that there is a global .npmrc file, so add it here as if it were in the repo.
-      def inferred_npmrc
+
+      def inferred_npmrc # rubocop:disable Metrics/PerceivedComplexity
         return @inferred_npmrc if defined?(@inferred_npmrc)
         return @inferred_npmrc = nil unless npmrc.nil? && package_lock
 
         known_registries = []
-        JSON.parse(package_lock.content).fetch("dependencies", {}).each do |_name, details|
-          resolved = details.fetch("resolved", "https://registry.npmjs.org")
+        JSON.parse(package_lock.content).fetch("dependencies", {}).each do |dependency_name, details|
+          resolved = details.fetch("resolved", DEFAULT_NPM_REGISTRY)
+
+          # skips values that are not strings. For example, when it is `false` as it will always fail to parse
+          next unless resolved.is_a?(String)
```

Comment:
This is not needed given the rescue below, but I think that it is a good addition

We could add a log in the rescue clause in the future. As I'd be interested to hear in what situations the parsing fails when the input is a string

Reply:
The `false` case, we do need it since I think it's due to an [upstream npm bug](https://github.com/npm/cli/issues/1138). But we should be able to remove it once we drop npm v6 support.

The other cases I guess they're there to handle corrupted lockfiles including malformed URIs. Not sure if that's worth supporting though.

In any case, all this seems unrelated to this fix so I'd prefer if we handled it separately.

So, my preference would be to leave this as is and let the diff in this PR and commit focus on the bug being fixed. You can extract this change to a separate PR and we can figure it out separately.

What do you think?

Reply:
That's fair. I updated the pull request removing that check. I'll send a follow up PR later today for the specific check

Thanks!

Reply:
I am behind on my notifications due to a head cold followed by food poisoning, so am not spending time to dig into all the details here, but @deivid-rodriguez wherever this npm6 workaround is, if it's not already let's try to make sure it's commented as such in the code? That way when we drop support for it, we can simply grep the codebase for "npm 6" to easily find these cases to delete them? Because if they're only discoverable via `git blame`, they'll get overlooked during the upgrade...

Reply:
Thanks for noticing @jeffwidman, I see now that this PR has actually removed the `# This can happen if resolved is false, for instance.` comment. That's my oversight since I was not expecting that comment to be removed and planned to handle it in the follow up PR including the change that I asked to be removed.

@yeikel Can you create a follow up PR to restore that comment and also link to the upstream issue that forces us to special case that?

Reply:
I will create one PR later today

Thanks!

Reply:
Submitted https://github.com/dependabot/dependabot-core/pull/7106